### PR TITLE
Support Pythia and Reorg Timm+OpenClip

### PR DESCRIPTION
**Pull request by Alexis-BX**

### CODE REVIEW

**Comment by Alexis-BX** [comment]
To review to ensure no loss by Dan and myself:
robin/model/multimodal_encoder/builder.py -> need to bring back local models
robin/model/multimodal_encoder/clip_encoder.py -> can you check Dan please? I'm quite sure we had changed this
robin/model/multimodal_encoder/open_clip.py -> need review
robin/train/train.py -> need review

**Comment by floatingsun** [comment]
Hi, I am not sure whether this should be putted into froniter branch as this is for training model in the machine that has no constrain for accessing the internet.

Regarding support local model, I can help set this branch up for frontier if we want to merge this!

File: robin/model/multimodal_encoder/clip_encoder.py [hunk 1]
```diff
@@ -29,7 +28,7 @@ def load_model(self):
     def feature_select(self, image_forward_outs):
         image_features = image_forward_outs.hidden_states[self.select_layer]
         if self.select_feature == 'patch':
-            image_features = image_features[:, 0:]
+            image_features = image_features[:, 1:]
```
Comment (daniel-z-kaplan) [inline robin/model/multimodal_encoder/clip_encoder.py:31]: We do not plan on supporting patch only - we support CLS only and patch + cls.

File: robin/model/multimodal_encoder/open_clip.py [hunk 2]
```diff
@@ -1,116 +1,68 @@
...
-            self.vision_tower, self.image_processor = create_model_from_pretrained(self.vision_tower_name)
-        
-
+        assert "/" in self.vision_tower_name, """
+            model name and pretrained must be split with / in open clip eg: ViT-B-16/laion2b_s34b_b88k"""
         
-        self.vision_tower = self.vision_tower.visual
-
-        if "timm" in self.vision_tower_name or name.lower() in TIMM_ON_OPENCLIP:
-            self.hidden_size = self.vision_tower.trunk.embed_dim
```
Comment (daniel-z-kaplan) [inline robin/model/multimodal_encoder/open_clip.py:47]: I am not seeing how this will work properly with Timm based openclip models if we remove this?

File: robin/model/multimodal_encoder/open_clip.py [hunk 3]
```diff
@@ -1,116 +1,68 @@
...
     # @torch.no_grad()
-    def forward(self, images): # shape, CLS shape, image tokens shape, and hidden size
-        
+    def forward(self, images):
         if type(images) is list:
             image_features = []
             for image in images:
-                cls_token, image_forward_out = self.vision_tower(image.to(device=self.device, dtype=self.dtype).unsqueeze(0))
+                image_forward_out = self.vision_tower(image.to(device=self.device, dtype=self.dtype).unsqueeze(0))
+                image_feature = self.feature_select(image_forward_out).to(image.dtype)
```
Comment (daniel-z-kaplan) [inline robin/model/multimodal_encoder/open_clip.py:50]: Is this code meant to work without the open_clip patching or something? It won't function with that.

File: robin/model/builder.py [hunk 4]
```diff
@@ -45,11 +45,24 @@ def load_pretrained_model(model_path, model_base, model_name, load_8bit=False, l
             # use_flash_attention_2 = True,
         )
     
-    token_num, tokem_dim = model.lm_head.out_features, model.lm_head.in_features
-    if model.lm_head.weight.shape[0] != token_num:
-        model.lm_head.weight = torch.nn.Parameter(torch.empty(token_num, tokem_dim, device=model.device, dtype=model.dtype))
-        model.model.embed_tokens.weight = torch.nn.Parameter(torch.empty(token_num, tokem_dim, device=model.device, dtype=model.dtype))
-
+    ## Below two paragraph of code are commented out because we don't actually
```
Comment (daniel-z-kaplan) [inline robin/model/builder.py]: As you mention here, this needs to be fixed. We don't train this with this *right now*, but if we do, the code will break.

File: robin/model/multimodal_encoder/builder.py [hunk 5]
```diff
@@ -1,30 +1,27 @@
...
-            print(f'Local model not handled in configs (might crash): {vision_tower}')
-            return OpenCLIPVisionTower(vision_tower, args=vision_tower_cfg, **kwargs)
+
+    if vision_tower in OPEN_CLIP_MODELS:
+        return OpenCLIPVisionTower(vision_tower, args=vision_tower_cfg, **kwargs)
+
+    elif vision_tower.lower().startswith("openai"):
+        return CLIPVisionTower(vision_tower, args=vision_tower_cfg, **kwargs)
+
+    elif vision_tower.lower().startswith("timm"):
```
Comment (daniel-z-kaplan) [inline robin/model/multimodal_encoder/builder.py]: We need a better way of doing this - let's think of one and add it. The 'timm' vision tower is special because it's for pure timm models, like Dino, as opposed to Open-Clip timm models, which are still through Timm, but some strange Open-Clip timm hybrid setup.

File: robin/model/multimodal_encoder/open_clip.py [hunk 6]
```diff
@@ -1,116 +1,67 @@
...
     def feature_select(self, image_forward_outs):
-        
-        assert False, ("not implemented")
-        
-        image_features = image_forward_outs.hidden_states[self.select_layer]
+        cls_token, patch = image_forward_outs
         if self.select_feature == 'patch':
-            image_features = image_features[:, 1:]
+            image_features = patch
         elif self.select_feature == 'cls_patch':
```
Comment (daniel-z-kaplan) [inline robin/model/multimodal_encoder/open_clip.py:39]: We plan to support cls only, and cls_patch. If we *Really* want we can support patch only also.

File: robin/model/multimodal_encoder/timm_vision.py [hunk 7]
```diff
@@ -10,65 +10,41 @@
...
-            
-            data_config = timm.data.resolve_model_data_config(self.vision_tower)
-            
-        self.hidden_size = self.vision_tower.num_features#Can also use embed_dim
-        
-        #might need to change n shit. might not.
-        transforms = timm.data.create_transform(**data_config, is_training=True)
-        
-        self.image_processor = transforms
-        self.vision_tower.requires_grad_(False)
```
Comment (daniel-z-kaplan) [inline robin/model/multimodal_encoder/timm_vision.py:61]: If we remove this from the encoder side, then we need to check it in train instead.

File: robin/model/multimodal_encoder/timm_vision.py [hunk 8]
```diff
@@ -10,65 +10,41 @@
...
-        
-        
     def feature_select(self, image_forward_outs):
-        
-        assert False, ("not implemented")
-        
-        image_features = image_forward_outs.hidden_states[self.select_layer]
+        image_features = image_forward_outs
         if self.select_feature == 'patch':
             image_features = image_features[:, 1:]
```
Comment (daniel-z-kaplan) [inline robin/model/multimodal_encoder/timm_vision.py]: This is a bug and needs to be fixed.

File: robin/model/multimodal_encoder/timm_vision.py [hunk 9]
```diff
@@ -79,46 +55,38 @@ def feature_select(self, image_forward_outs):
...
-                image_features.append(image_forward_out)
-        else:#This should always be unsqueezed, if we have multiple items just stack them before this
-            cls_token, image_features = self.vision_tower(images.to(device=self.device, dtype=self.dtype))
-        cls_token = cls_token.unsqueeze(1)
-        image_features = torch.cat((cls_token, image_features), dim=1)
+                image_feature = self.vision_tower.forward_features(image.to(device=self.device, dtype=self.dtype))
+                image_feature = self.feature_select(image_feature).to(image.dtype)
+                image_features.append(image_feature)
+        else:
+            # This incl cls token
```
Comment (daniel-z-kaplan) [inline robin/model/multimodal_encoder/timm_vision.py]: How does this handle the cls token then...?

File: robin/train/train.py [hunk 10]
```diff
@@ -978,9 +1041,8 @@ def make_inputs_require_grad(module, input, output):
                     args=training_args,
                     **data_module)
 
-
     if training_args.finetune_ve:
-        for name, param in model.base_model.model.model.vision_tower.named_parameters():#This is required for lora, and training without lora will not work on this line.
+        for name, param in model.base_model.get_vision_tower().named_parameters():#This is required for lora, and training without lora will not work on this line.
```
Comment (daniel-z-kaplan) [inline robin/train/train.py]: Does this work for all model types?

File: robin/model/multimodal_encoder/clip_encoder.py (hunk 1 above)
Comment (floatingsun) [reply robin/model/multimodal_encoder/clip_encoder.py:31]: I will add the cls only later.

File: robin/model/multimodal_encoder/open_clip.py (hunk 2 above)
Comment (floatingsun) [reply robin/model/multimodal_encoder/open_clip.py:47]: add test for `ViT-SO400M-14-SigLIP-384`

File: robin/model/multimodal_encoder/builder.py (hunk 5 above)
Comment (floatingsun) [reply robin/model/multimodal_encoder/builder.py]: Add new arg called `vision type`, parse this arg to select the visiontower class.
and create a outside sheet to map each vision tower name to its type.

File: robin/model/multimodal_encoder/open_clip.py (hunk 2 above)
Comment (floatingsun) [reply robin/model/multimodal_encoder/open_clip.py:47]: Hi, i test at here, but one wired thing I found is that forward_features  return without cls token, is this because that siglip don't have such cls token, or something else?

 https://github.com/AGI-Collective/Robin/commit/431ac9c1bb7dc5d41731dfeacaf087d66e5c8dca

File: robin/model/multimodal_encoder/builder.py (hunk 5 above)
Comment (floatingsun) [reply robin/model/multimodal_encoder/builder.py]: Pls check this commit for add `vision_tower_type`
https://github.com/AGI-Collective/Robin/commit/e04d449459389f566433ff88344380ad06a5a8ca
